### PR TITLE
[ruby] Make `GRPC::DeadlineExceeded` report properly

### LINKED ISSUE #33283: ruby: Server stream reports "grpc_call_start_batch failed with outstanding read or write present (code=8)" instead of DeadlineExceeded
### What version of gRPC and what language are you using?

grpc v1.55.0

### What operating system (Linux, Windows,...) and version?

Linux Ubuntu 20.04

### What runtime / compiler are you using (e.g. python version or version of gcc)

Ruby 3.1.4, 3.0.6

### What did you do?

I reproduced the error with this example:

```proto
syntax = "proto3";

option ruby_package = "hello_pb";

service HelloService {
  rpc SayHello (HelloRequest) returns (stream HelloResponse) {}
}

message HelloRequest {
  string name = 1;
}

message HelloResponse {
  repeated string messages = 1;
}
```

```ruby
require 'grpc'
require_relative 'hello_services_pb'

class HelloServer < HelloPb::HelloService::Service
  def say_hello(request, _call)
    sleep 5
  end
end

def main
  server = GRPC::RpcServer.new
  server.add_http2_port('0.0.0.0:50051', :this_port_is_insecure)
  server.handle(HelloServer)
  server.run_till_terminated_or_interrupted([1, 'int', 'SIGQUIT'])
end

main
```

```ruby
require 'grpc'
require_relative 'hello_services_pb'

def main
  stub = HelloPb::HelloService::Stub.new('127.0.0.1:50051', :this_channel_is_insecure)
  request = HelloPb::HelloRequest.new(name: 'World')
  response = stub.say_hello(request, deadline: Time.now + 1)

  response.each do |response|
    puts "response: #{response.messages}"
  end
end

main
```

### What did you expect to see?

With grpc v1.45.0, I see `GRPC::DeadlineExceeded`:

```
$ ruby -I. client.rb
/home/stanhu/.asdf/installs/ruby/3.0.6/lib/ruby/gems/3.0.0/gems/grpc-1.42.0-x86_64-linux/src/ruby/lib/grpc/generic/active_call.rb:29:in `check_status': 4:Deadline Exceeded. debug_error_string:{"created":"@1685516791.527016274","description":"Deadline Exceeded","file":"src/core/ext/filters/deadline/deadline_filter.cc","file_line":81,"grpc_status":4} (GRPC::DeadlineExceeded)
        from /home/stanhu/.asdf/installs/ruby/3.0.6/lib/ruby/gems/3.0.0/gems/grpc-1.42.0-x86_64-linux/src/ruby/lib/grpc/generic/active_call.rb:180:in `attach_status_results_and_complete_call'
        from /home/stanhu/.asdf/installs/ruby/3.0.6/lib/ruby/gems/3.0.0/gems/grpc-1.42.0-x86_64-linux/src/ruby/lib/grpc/generic/active_call.rb:169:in `receive_and_check_status'
        from /home/stanhu/.asdf/installs/ruby/3.0.6/lib/ruby/gems/3.0.0/gems/grpc-1.42.0-x86_64-linux/src/ruby/lib/grpc/generic/active_call.rb:337:in `each_remote_read_then_finish'
        from client.rb:9:in `each'
        from client.rb:9:in `main'
        from client.rb:14:in `<main>'
```

### What did you see instead?

Starting with grpc v1.46.2 and up, I see `grpc_call_start_batch failed with outstanding read or write present (code=8)`:

```
$ ruby -I. client.rb
/home/stanhu/.asdf/installs/ruby/3.0.6/lib/ruby/gems/3.0.0/gems/grpc-1.46.3-x86_64-linux/src/ruby/lib/grpc/generic/active_call.rb:169:in `run_batch': grpc_call_start_batch failed with outstanding read or write present (code=8) (GRPC::Core::CallError)
        from /home/stanhu/.asdf/installs/ruby/3.0.6/lib/ruby/gems/3.0.0/gems/grpc-1.46.3-x86_64-linux/src/ruby/lib/grpc/generic/active_call.rb:169:in `receive_and_check_status'
        from /home/stanhu/.asdf/installs/ruby/3.0.6/lib/ruby/gems/3.0.0/gems/grpc-1.46.3-x86_64-linux/src/ruby/lib/grpc/generic/active_call.rb:343:in `each_remote_read_then_finish'
        from client.rb:9:in `each'
        from client.rb:9:in `main'
        from client.rb:14:in `<main>'
```

### PR DESCRIPTION
**Pull request by stanhu**
In grpc v1.46.2 and later versions, #29155 caused the Ruby client to return `GRPC::Core::CallError`, a non-standard error, with a message: `grpc_call_start_batch failed with outstanding read or write present (code=8)`. However, the actual error should have been `GRPC::DeadlineExceeded`. This occurred because when `GRPC::DeadlineExceeded` is raised, the `@metadata_received` flag doesn't get flipped. When `receive_and_check_status` runs to determine
the error, the `RECV_INITIAL_METADATA` is erroneously sent again.

To avoid this, flip the flag in an `ensure` block whenever the `RECV_INITIAL_METADATA` op is set.

Closes #33283

### CODE REVIEW

**Comment by stanhu** [comment]
@apolcyn What do you think of this fix?

File: src/ruby/lib/grpc/generic/active_call.rb
Comment (alto-ruby) [inline src/ruby/lib/grpc/generic/active_call.rb:178]: This becomes the opposite that of the original report that if `run_batch` throws before sent RECV_INITIAL_METADATA, we are setting metadata_received anyways.

Ideally we need to c-core to propagate the error from [completion queue](https://github.com/grpc/grpc/blob/v1.56.0/src/core/lib/surface/completion_queue.cc#L773) so here we know whether to metadata_received or not.

Given  that's a fairly large change in c-core, I incline to accept this to unblock the customer as I mentioned in the other comment in the issue, missing a RECV_INITIAL_METADATA doesn't really cause problems (for now at least).

@apolcyn Do you have any other concerns?

**Review by alto-ruby — APPROVED** [review]

**Comment by faustind** [comment]
🙌🏾 

Thanks to all who are working to resolve https://github.com/grpc/grpc/issues/33283. 

With increased traffic between our gRPC services, the frequency of this non-standard error being reported has also increased.

This PR was approved by @alto-ruby approximately 2 months ago. However, it seems there has been no further activity since then.

We look forward to this PR being merged.

**Comment by stanhu** [comment]
I'm seeing a conflict with #34016, which seems like it squelches the error? I'm not sure that's what we want; we want to see `DeadlineExceeded`.

**Comment by xlgmokha** [comment]
@alto-ruby is there anything that we can do about the concern raised [here](https://github.com/grpc/grpc/pull/33565#issuecomment-1817000994). I'd love to see this patch released.

**Comment by alto-ruby** [comment]
@apolcyn Do you have any concerns to tis PR?

**Comment by alto-ruby** [comment]
@apolcyn 
Hi Alex, happy new year! 
Could we get your approval of this PR? 
I'll wait one more week to see if you have any concerns, if not, I'll go ahead and merge it. Sounds OK?

**Comment by HannahShiSFB** [comment]
@sampajano Hi Eryu, could you help me check the "import/copybara Pending" status? Thanks a lot!

**Comment by sampajano** [comment]
@HannahShiSFB Thanks for checking! I've approved the internal merge request and will find one more reviewer to approve too (2 is required)!

**Comment by sampajano** [comment]
@alto-ruby @HannahShiSFB There's some test failures. Can you work on resolving them? Thanks! :)

**Comment by stanhu** [comment]
The test failures look unrelated to this change. I rebased `master`.

**Comment by sampajano** [comment]
@stanhu Thanks for checking and rebasing! Running the tests again now! 😃

**Comment by stanhu** [comment]
Test failures look unrelated again. 🤷

**Comment by sampajano** [comment]
ok.. that's unfortunate.. can you maybe rebase again? 😅

**Comment by stanhu** [comment]
@sampajano Rebased again. Can you launch a CI run?

**Comment by sampajano** [comment]
Thank you! Done! Fingers crossed.. 😄

**Comment by stanhu** [comment]
So close! Everything passed except for https://source.cloud.google.com/results/invocations/eeb9b8bd-12a5-4887-a504-47b85ac55c06/targets, but this seems like a flaky test.

**Comment by sampajano** [comment]
Looks like it! Lemme run it again! 😄

**Comment by stanhu** [comment]
@sampajano Could we get this merged?

**Comment by sampajano** [comment]
@stanhu I'm sorry for the delays.. But it seems like there's another unrelated test failure.. 😅

I'm checking with the team how to get it merged in spite of the error..

But in the meantime, do you mind rebasing yet again and i'll run the test again..?

Apologize again for the trouble..

**Comment by sampajano** [comment]
@stanhu I manually re-imported the PR and it worked!! Thanks so much for your contribution here again! And sorry for the delays! 😃

**Comment by stanhu** [comment]
@sampajano Thank you!

**Comment by sampajano** [comment]
@stanhu you're most welcome! Thank you again for your contribution! 😊

**Pull request closed without merging**
